### ISSUE TITLE
client.setQueryData is so slow

### ISSUE THREAD

**Issue by Evertt** [DamianOsipiuk/vue-query#208]
I have this code:

```ts
const client = useQueryClient()
client.setQueryData(key, (old: any[]) => old.map(
    item => item.id === variables.id ? optimistic : item
))
```

Where I basically optimistically update an entry in a list before a mutation function is run. Now sometimes the list is 3000 entries long. And then this line of code takes 18 seconds to run on my machine. That's crazy slow. And I found out, it's not even because of the map function. When I do `client.setQueryData(key, old => old)`, it still takes around 18 seconds.

Now the fact that it takes 18 seconds for `client.setQueryData()` to complete isn't the only problem. The other problem is that during those 18 seconds, the UI of the site is completely unresponsive.

Why is this so slow? And can I do something to make this more optimized? Or at the very least keep the UI responsive while this is doing its work.

**Comment by DamianOsipiuk** [comment]
Hi,
So I have few additional questions here that might help to narrow it down:
- is it happening in both v1 and v2?
- is is happening in https://github.com/DamianOsipiuk/vue-query/releases/tag/v1.19.3? (1.20 introduced wrapper around QueryClient which might be the culprit here)
- could you check if disabling `stucturalSharing` helps? 
  https://vue-query.vercel.app/#/guides/important-defaults
  ```ts
  const query = useQuery(["posts"], fetcher, {
    structuralSharing: false,
  });
  ```

**Comment by Evertt** [comment]
Okay well based on your questions I started some experiments and funnily enough what solved it for me was to update the package from `v1.19.1` to `1.25.0`. The time to run that function went down from 18 seconds to 600 ms.

The `structuralSharing: false` didn't make a difference, on either version. Also I'm hesitant to try v2, because I'm afraid the breaking changes may mean it'll take me quite some time to get everything to work again.

**Comment by Evertt** [comment]
Hi, I'm back.

So like I said in the last message, updating to the most recent 1.x version did make `client.setQueryData()` *much* more performant.

And yet, it doesn't seem to be performant enough. Or at least, we do still have problems with when we're updating a single row, the UI becomes very laggy until the update is fully complete.

I'm kind of wondering if it would be possible to move all `vue-query`'s work onto a Web Worker. I think it should be possible in theory, but I don't think that I can do that by myself. Because I think the implementation of that would need to start in the core of `vue-query`.

What do you think? And also, you mentioned that I could also upgrade to v2 and I was hesitant to do that. But how confident are you that v2 could possibly solve this issue for me? If you're very confident about it then I'm willing to try to upgrade.

**Comment by DamianOsipiuk** [comment]
So v2 would have some performance benefits, but from your description it seems that the issue might still persist.

Few ideas that i have is that:
1. you are doing some expensive computations / shape modifications in `computed` / `watch` / `select` that is triggered immediately after query updates
2. you are optimistically updating a cache item that has millions of items
3. your rendering logic is not stable and is replacing loads of items on each rendering loop?

Unfortunately there is so much that i can think of without the actual reproduction example.
What you could do is to perf test it more to nail down on what exact function it spends the most time when optimistically updating. Is it really vue-query internals or something in user land.

**Comment by Evertt** [comment]
Hey man,

Thanks for replying, I really appreciate it. So I did a performance test. Not with chrome's devtools, because I don't really understand the performance tab there. 😅  But by writing `console.time() ... console.timeEnd()` everywhere and I found out that indeed all the performance problems are on my side.

One part is that I set my project up in such a way that when a query fetches their endpoint, it also triggers every other query that has some relation to the first query (which may in turn also trigger other queries for their relations), etc etc. So that at the end, what I get back is an object where every `user_id`, `order_id`, etc, is just converted into an actual `User` object and `Order` object.

But of course this also means that when I update the cache of one query, it triggers a cascade of all these mappings to be done again. And yeah some of my query caches contain thousands of rows. Not millions, but still.

But actually the biggest performance bottleneck that I have is that I'm saving the full query cache, of all queries, into indexedDB. And that gets triggered on every change of anything. I do have a throttle on that of 500ms, but apparently it's still slowing things down a lot. And because you get an error if you have multiple tabs open that are trying to access the same indexedDB database, I made it so that every time it wants to save new data in indexedDB, it will open a new connection to the store, store the data, and then close the connection. Which is much slower than if I would just keep the connection open.

What I'm gonna try to do, is to move all of that logic, into a web worker. See if that at least helps to not make the UI freeze. I'm also wondering if I should try to do all the mapping of query data relations in a web worker too. If that's even feasible. 🤔 

Anyway, that's my update. If you have any further advice, I'd love to hear it.
